File: agent/src/GaltekClassroom.Agent.Service/Master/MasterWindowsBinding.cs
using System.Text.Json.Serialization;
using GaltekClassroom.Agent.Shared;

namespace GaltekClassroom.Agent.Service.Master;

public sealed record MasterWindowsBinding
{
    [JsonPropertyName("schemaVersion")]
    [JsonPropertyOrder(0)]
    public int SchemaVersion { get; init; } = MasterBindingConstants.SchemaVersion;

    [JsonPropertyName("installationId")]
    [JsonPropertyOrder(1)]
    public Guid InstallationId { get; init; }

    [JsonPropertyName("windowsSid")]
    [JsonPropertyOrder(2)]
    public string WindowsSid { get; init; } = string.Empty;

    [JsonPropertyName("accountDisplayName")]
    [JsonPropertyOrder(3)]
    public string AccountDisplayName { get; init; } = string.Empty;

    [JsonPropertyName("boundAtUtc")]
    [JsonPropertyOrder(4)]
    public DateTimeOffset BoundAtUtc { get; init; }

    public static MasterWindowsBinding Create(
        Guid installationId,
        string windowsSid,
        string accountDisplayName,
        DateTimeOffset boundAtUtc)
    {
        return new MasterWindowsBinding
        {
            SchemaVersion = MasterBindingConstants.SchemaVersion,
            InstallationId = installationId,
            WindowsSid = windowsSid,
            AccountDisplayName = accountDisplayName,
            BoundAtUtc = boundAtUtc.ToUniversalTime()
        };
    }
}

File: master-backend/src/main/java/com/galtek/classroom/master/MasterWindowsBinding.java
package com.galtek.classroom.master;

import static com.galtek.classroom.domain.DomainChecks.requireNonBlank;
import static com.galtek.classroom.domain.DomainChecks.requireNonNull;

import java.time.OffsetDateTime;

public record MasterWindowsBinding(
        String installationId,
        String windowsSid,
        String accountDisplayName,
        OffsetDateTime boundAtUtc) {

    public MasterWindowsBinding {
        installationId = requireNonBlank(installationId, "installationId");
        windowsSid = requireNonBlank(windowsSid, "windowsSid");
        accountDisplayName = requireNonBlank(accountDisplayName, "accountDisplayName");
        requireNonNull(boundAtUtc, "boundAtUtc");
    }
}
